Source/DafnyCore/Verifier/Translator.cs

```diff
@@ -125,6 +126,55 @@ public class TranslatorFlags {
     readonly Dictionary<string, Bpl.Constant> fieldConstants = new Dictionary<string, Constant>();
     readonly Dictionary<string, Bpl.Constant> tytagConstants = new Dictionary<string, Constant>();
 
+    public class ProofDependencyManager {
+      // proof dependency tracking state
+      public Dictionary<string, ProofDependency> ProofDependenciesById { get; } = new();
+      private UInt64 proofDependencyIdCount = 0;
+
+      public string GetProofDependencyId(ProofDependency dep) {
+        var idString = $"id{proofDependencyIdCount}";
+        ProofDependenciesById[idString] = dep;
+        proofDependencyIdCount++;
+        return idString;
+      }
+
+      public void AddProofDependencyId(ICarriesAttributes boogieNode, IToken tok, ProofDependency dep) {
+        var idString = GetProofDependencyId(dep);
+        boogieNode.Attributes =
+          new QKeyValue(tok, "id", new List<object>() { idString }, boogieNode.Attributes);
```

Comment:
Why are we using a string `idString` instead of directly storing `dep` in the attribute, so you won't need the `ProofDependencyManager`? 

If the purpose is for the Boogie code to be printable and for a person to be able to map that print back to the original Dafny code, then I think you'll have to base the ids on Dafny source locations instead of generating them, and then you also wouldn't need a `ProofDependencyManager`.

Reply:
The reason for this is that Boogie needs to have a string to send to Z3 -- that's the only way to specify the label on an expression you want to track for proof dependencies. Then Z3 returns a set of strings indicating which pieces it used. The notion of proof dependency that we have in Dafny is Dafny-centric, so it wouldn't make sense to have the `ProofDependency` class in Boogie. It's possible that we could have an additional layer of abstraction in the Boogie implementation, so we'd have 1) names as strings, 2) structured Boogie conceptions of those strings, and 3) structured Dafny conceptions that build on the Boogie ones. But I'm not sure that that additional layer is really worth it. Ultimately, the parsing we do needs to happen _somewhere_. It's just a question of whether it's in Boogie or Dafny.

Reply:
> The reason for this is that Boogie needs to have a string to send to Z3

Why not do the id generation with AST node mapping in Boogie? Seems better to let Boogie encapsulate SMT generation. Then you also don't need the method `GetFullIdDependency` in Dafny and references to the coverage related string constants. `VCResult.coveredElements` would then have type `IEnumerable<Absy>` instead of `IEnumerable<string>`.

Reply:
Although on the surface it seems like an `Absy` would encapsulate everything that we need, in practice this isn't the case because of how Boogie transforms the original program and how particular nodes are used in different ways in a proof (such as how a requires clause is an assumption when proving a definition itself, but a proof goal when invoking that definition). So, we'd need something like `ProofDependency` in Boogie, but it couldn't be quite the same thing, because those program constructs don't map cleanly back from Boogie to Dafny. So we'd need both a BoogeProofDependency and a DafnyProofDependency, and I don't think that extra layer of indirection is worthwhile at the moment.

Reply:
> Although on the surface it seems like an Absy would encapsulate everything that we need, in practice this isn't the case because of how Boogie transforms the original program and how particular nodes are used in different ways in a proof (such as how a requires clause is an assumption when proving a definition itself, but a proof goal when invoking that definition)

I don't see that, since Dafny attaches each id to a unique `Absy` instance. Regarding your example, there a separate `Absy` for each pair of call and requires clause, and a separate one for each requires clause.

---

This PR lets Dafny attach ids to `Absy`s, while making an id -> `proofDependency` mapping, then it takes the ids returned by Boogie that gets them from the unsat core, and maps them back to `proofDependency`s.

Alternatively, you can have Boogie attach ids to the `Absy`s, map the ids coming out of the unsat core back to `Absy`s, and then have Dafny map the Absy's back to `proofDependency`s, so Dafny doesn't have to deal with the ids.

Reply:
Dafny can't keep track of an Absy -> ProofDependency map because Boogie introduces new Absy nodes during internal transformations that Dafny knows nothing about. Boogie would end up returning Absy nodes that Dafny has never seen, and therefore can't be in the map Dafny created.

Reply:
> Dafny can't keep track of an Absy -> ProofDependency map

In what I described, it does not. Dafny does not keep track of anything. It retrieves an `IEnumerable<Absy>` from `VCResult.coveredElements` and those `Absy` contain `IToken`s with all the description that `ProofDependency` currently provides.

Reply:
But `IToken` values don't (and I believe shouldn't) track all of the information that `ProofDependency` values do.

Reply:
In particular, at least one of the `ProofDependency` types, and likely more in the future, keep the original _Dafny_ AST node that gave rise the dependency. That's only lightly used at the moment, but will need to be used more heavily in the upcoming changes I have planned.

Reply:
It is also important to be able to print the Boogie files, as you suggest above, and encoding IDs in those that refer simply to Dafny source locations won't be enough, because Dafny source locations overlap too much. A single expression can result in multiple proof dependencies.